File: DynamicPrograming/python/sumPossible.py
from functools import lru_cache
from typing import List


def sumPossible(amount: int, numbers: List[int]) -> bool:
    @lru_cache(maxsize=None)
    def dfs(remaining: int) -> bool:
        if remaining == 0:
            return True
        if remaining < 0:
            return False
        for num in numbers:
            if dfs(remaining - num):
                return True
        return False

    return dfs(amount)

File: DynamicPrograming/go/sum_possible.go
package dynamicprogramming

// SumPossible determines if amount can be formed using numbers.
func SumPossible(amount int, numbers []int) bool {
	memo := make(map[int]bool)

	var dfs func(int) bool
	dfs = func(remaining int) bool {
		if remaining == 0 {
			return true
		}
		if remaining < 0 {
			return false
		}
		if val, ok := memo[remaining]; ok {
			return val
		}
		for _, num := range numbers {
			if dfs(remaining - num) {
				memo[remaining] = true
				return true
			}
		}
		memo[remaining] = false
		return false
	}

	return dfs(amount)
}
